### DISCUSSION TITLE
20.1-0.14-v2 Update

### DISCUSSION THREAD

**Discussion by BeansGalaxy** [BeansGalaxy/Beans-Backpacks-2#69] [Feedback]
This update adds the Winged and Ender Backpacks, but I know the idea's aren't fully formed just yet. So far I have these issues with each,

## Ender Backpacks
- Ender Backpacks cant be taken off while items are inside with the current controls. I like the current controls and I would like to keep it the same with other backpacks but don't know how allow the player to take off the backpack. My best idea is letting players take it off with a shift click.

## Winged Backpacks
- Their durability replaced the usual backpack fullness bar in the inventory, I'll look into adding a second durability bar and I don't know if its necessary to show the durability in the menu while it's placed but any idea's how to make it look?
- Also I want to display the elytra wings while it's placed but it would clip into the surrounding blocks if I did. Is that a problem or should I implement it?

Also leave any other suggestions or comments you might have for this update here, I'm super excited to hear your feedback.

**Comment by LegendarySpy** [comment]
I think the elytra's on the backpack clipping would be fine for the most part, you could also try like shrinking them / folding them up to the size of the backpack when the backpack is on the ground. Otherwise nice mod, its a great addition to my pack 👍

Also. is there a way to fix this weird initial soft-crash?
![image](https://github.com/BeansGalaxy/Beans-Backpacks-2/assets/63053684/eb62a9aa-e078-4a51-ae4e-6d9a3597bd31)

**Reply by BeansGalaxy** [reply]
Attach a zip file of the world if it continues to happen. I had the same thing happen while coding a few times but I could never re-create it and don't know how it fixed itself.

**Comment by Alvin21Bon** [comment]
This is a great update and I love the direction the mod is going in so far, good work on the game design as always. 

# Bugs

**Note, I am testing this while using KubeJS to add the datapack configuration, so I am unsure if this exists normally yet KubeJS should work the exact same as datapacks. Do also note that on a vanilla instance with just this mod installed, the elytra bug still exists.** If an item is in `disables_back_slot` both are still able to be worn if first the backslot is occupied, then the chestplate item is equipped. Also, if an item is in `disable_chestplate` it can still be put into the chestplate by any means other than dispenser equipping. Last, when an item is in `elytra_items` it can still be equipped at the same time as a backpack by first wearing a backpack, then shift clicking the item into the chestplate spot.

For another configuration bug **(that might be KubeJS specific,)** every configuration file that is made in the datapack overwrites what is already present in the mod. This means that making an empty `blacklist_items` file in a databack actually just removes every shulker box that is already in `blacklist_items`. 

# Ender Backpack
The ender backpacks are a good addition and is very good for offering functionality that is different from shulker boxes when paired with a hopper. Overall, very nice!

One thing I notice though if I understand how the ender backpacks work, is that there is no way to unbind all ender backpacks in the world from your name. One result of this is that someone nefarious could steal your ender backpack or simply kill you while you are wearing one. This means then that ender backpacks are no longer usable for you since this other player will have your binded backpack, meaning they can just funnel whatever items you get with a hopper permanently. I like how the binded backpacks system works to differentiate it from ender chests, but I do feel this can be a problem in servers.  

Another thing is that an ender backpack hooked up to a hopper obviously does not work when the chunk is unloaded. I know there is no easy solution to this, but I feel the use case of using them while exploring or caving is limited due to chunks unloading. Hopefully a way to solve this is even possible, I am not familiar with modding this game.

# Winged Backpacks
I love that this idea was able to get past the planning phase and actually work well when executed. As for the wings clipping if rendered while placed down, I don't think this would be a problem. Even vanilla blocks have models that clip into blocks, like the lectern. Maybe have the elytra fold upwards like how it is while flying so it is all visible when placed down. 

I also feel like the winged backpack itself should have the elytra somehow incorporated in its item model. I found it confusing that the elytra spawns on my back when the backpack is equipped. A pretty out there and weird solution to solve some of these issues might be to instead have the elytra *inside* of the backpack. So, the model will maybe have the elytra peeking out the flap and when used will fully open up the backpack and have the elytra fold out of the backpack. I know this would probably be difficult lol, just a weird idea. 

Also, an idea for the durability of the elytra would be to include it in the backpack info. I feel that having two durability bars visible would be a bit confusing unless you plan to put an indicator for which bar is which.

**Reply by Alvin21Bon** [reply]
Forgot to mention about taking off the ender chest: I think that just accessing the items when it is placed down would be good, and then removing the backpack by just clicking it out of the backslot. Although, this would be inconsistent with all other backpacks which allows for clicking items out of it, I tend to not really use this method that much anyway since I can just place the backpack down fast.

Also, was wondering what your stance is on trimming leather backpacks since leather armor is able to be trimmed in game.

**Reply by BeansGalaxy** [reply]
A lot of good ideas in here.

## Bugs

I'll have to test out `disables_back_slot` & `disable_chestplate` again. I must've forgot to re-implement the code disabling them when I added `elytra_items`. As for dispensers, that's an oversight and I'll fix that as well as right clicking since I probably didn't change that either. 

I'll also have to watch what happens to the data when you try to add items to the lists. 

## Ender Backpacks

I might add coordinates to the Ender Backpack Menu if F3 is on and you are the owner. Not too sure how to add gui elements but it's a better way than just giving up on all the backpacks you have. Also going to make the backpacks forget their owner upon death so it would be as easy.

I thought about chunk loading Ender Backpacks until I realized there's already an easy way with vanilla nether portals so I scrapped the idea. Also its probably not a great idea to let a toxic player carry around 36+ chunk loaders in their back pocket.

Taking off the Ender Backpack is another good idea, I was thinking pressing shift + hotkey could move the item off but your idea seems easier to understand.

## Winged Backpacks

I'm thinking the wings would come straight out horizontally when it's hung and doesn't display when placed on the top/bottom of a block. I'm re-working the Backpack Entity code to allow for more customization in the future so I like my idea because I can actually implement it for now. It's still up in the air. (punny)

The item's texture right now is literally the leather backpack I have no clue how to fit wings and a backpack into 16x16 pixels so that's what we have for now. I need to re-do the ender backpack too. 

Also I think you get the idea that the bar associated with the durability because it's color and context. I'm probably going to leave it as is for now since Winged/Ender Backpacks are sort of the advanced backpacks and I don't think you'd get them until you're accustomed with the previous backpacks.

**Reply by Alvin21Bon** [reply]
I actually did not think about vanilla chunk loaders for the ender backpacks and that makes a lot more sense. It falls in line with how all other persistent contraption issues are dealt with in the vanilla game. As for the ender backpack stealing, would it be good also if possibly only the owner of an ender backpack is able to change the name of it through an anvil? Then the F3 menu or whatever indicator could show the coordinates of the backpacks that are currently loaded and the name associated with it. I feel this could help for making your own backpack and sharing it with friends, so you can name it accordingly and if something goes wrong like someone stealing from your ender backpack then you could track down which one it is.

**Reply by BeansGalaxy** [reply]
Worked out a lot of security things with ender backpacks like: display's their owner's name, shows chords if you're the owner and F3 is on *(only placed backpacks for now)*, unbinds on death, and you cant open it if the player is offline. *(those last too I'm working on adding gamerules to let those be disabled)*.

Also for `!namespace:id` overriding the mod's files. It's because the files are setup wrong. instead of `data/beansbackpacks/modify` you should use `data/[literally anything else than beansbackpacks]/modify`. Misunderstanding on my part, I'll update the wiki. Also I have now a template you can start with [here](https://github.com/BeansGalaxy/Beans-Backpacks-2/tree/20.1/assets/resources/Modify%20Data%20Template).

**Reply by BeansGalaxy** [reply]
I like the idea of naming them in an anvil too, I'll add that when I add tracking players wearing your ender backpack.
